pageserver/src/tenant_mgr.rs

```diff
@@ -3,52 +3,127 @@
 }
 
-pub fn init(conf: &'static PageServerConf) {
-    let mut m = REPOSITORY.lock().unwrap();
+fn access_repository() -> MutexGuard<'static, HashMap<ZTenantId, Arc<dyn Repository>>> {
+    REPOSITORY.lock().unwrap()
+}
+
+pub fn init(
+    conf: &'static PageServerConf,
+    storage_accessor: Option<StorageAccessor>,
+) -> Option<JoinHandle<anyhow::Result<()>>> {
+    let (storage_uploader, handle) = match storage_accessor {
+        Some(StorageAccessor {
+            storage_uploader,
+            download_signal,
+        }) => {
+            let storage_uploader = Arc::new(storage_uploader);
+            let handle = launch_download_listener_thread(
+                conf,
+                Arc::clone(&storage_uploader),
+                download_signal,
+            );
+            (Some(storage_uploader), Some(handle))
+        }
+        None => (None, None),
+    };
 
+    let mut m = access_repository();
     for dir_entry in fs::read_dir(conf.tenants_path()).unwrap() {
         let tenantid =
             ZTenantId::from_str(dir_entry.unwrap().file_name().to_str().unwrap()).unwrap();
+        let repo = init_repo(conf, storage_uploader.as_ref().map(Arc::clone), tenantid);
+        info!("initialized storage for tenant: {}", &tenantid);
+        m.insert(tenantid, repo);
+    }
 
-        // Set up a WAL redo manager, for applying WAL records.
-        let walredo_mgr = PostgresRedoManager::new(conf, tenantid);
+    handle
+}
 
-        // Set up an object repository, for actual data storage.
-        let repo = Arc::new(LayeredRepository::new(
-            conf,
-            Arc::new(walredo_mgr),
-            tenantid,
-        ));
-        LayeredRepository::launch_checkpointer_thread(conf, repo.clone());
-        LayeredRepository::launch_gc_thread(conf, repo.clone());
+fn init_repo(
+    conf: &'static PageServerConf,
+    storage_uploader: Option<Arc<StorageUploader>>,
+    tenant_id: ZTenantId,
+) -> Arc<LayeredRepository> {
+    // Set up a WAL redo manager, for applying WAL records.
+    let walredo_mgr = PostgresRedoManager::new(conf, tenant_id);
 
-        info!("initialized storage for tenant: {}", &tenantid);
-        m.insert(tenantid, repo);
+    // Set up an object repository, for actual data storage.
+    let repo = Arc::new(LayeredRepository::new(
+        conf,
+        Arc::new(walredo_mgr),
+        tenant_id,
+        storage_uploader,
+    ));
+    LayeredRepository::launch_checkpointer_thread(conf, repo.clone());
+    LayeredRepository::launch_gc_thread(conf, repo.clone());
+    repo
+}
+
+fn launch_download_listener_thread(
+    conf: &'static PageServerConf,
+    storage_uploader: Arc<StorageUploader>,
+    download_receiver: Receiver<(ZTenantId, ZTimelineId)>,
+) -> JoinHandle<anyhow::Result<()>> {
+    std::thread::Builder::new()
+        .name("Relish storage download listener thread".into())
+        .spawn(move || {
+            while let Ok((tenant_id, timeline_id)) = download_receiver.recv() {
```

Comment:
Second thread, listening for the sender events from the main upload/download loop.
Due to this condition and according to the docs, the loop stops the moment the sender is dropped.